### ISSUE TITLE
Performance of Geod.fwd for large numpy arrays as input.

### ISSUE THREAD

**Issue by thor-andersen** [jswhit/pyproj#106]
Hi. 
I am a big fan of the pyproj implementation.
My issue is more a question rather than an issue. 

I am working on a project where i need to make a large number of points along a route consisting of great circle lines. 

I am using g.fwd to create the sample points using the initial azimuth and an array of partial distances from the waypoints to create the points. The code is fast, but I need between 25,000 and 100,000 points along a given route and the creation (g.fwd) of the points ends up being a bottleneck. 

Do anyone have a suggestion to increase performance? 
Can I somehow trick npts to take an array of waypoints as input? looping through npts for each part of the route is significantly slower.

Multiprocessing is not an option as the main program is already running calculations like this in parallel.

A simple example is included below:
```
#Finding the distances and directions for traveled route legs    
import numpy as np
from pyproj import Geod
g = Geod(ellps='clrk66') # Geod definition

#Definition of waypoints
lon0 = np.array([-44., -42.25797664, -40.50913661, -37.,
                 -33.49086339, -31.74202336, -30.])         
lat0 = np.array([ 56.5,  56.58622486,  56.64795118,  56.69741501,
                 56.64795118,  56.58622486,  56.5])
#Calculating forward and back azimuths and distances between waypoints
az12,az21,distances = g.inv(lon0[0:-1],lat0[0:-1],lon0[1:],lat0[1:])

#1000 samplepoints between each waypoint
points = np.ones(np.size(lon0)-1).astype(int) * 100

#Creating distance factor from a waypoint to a given samplepoint
linspace_array = np.linspace(0, 1. - 1./points[0],points[0])
#Repeating array to create distance factor for each samplepoint between waypoints
points_space = np.tile(linspace_array,len(points))
#Creating the longitude start array, excluding the end point
lon0_start_array = np.repeat(lon0[0:-1],points)
#Creating the latitude start array, excluding the end point
lat0_start_array = np.repeat(lat0[0:-1],points)

#Creating the array of dirrections - repeating the direction for 
#the number of points you want between two points
#Creating arrys used to find the sampling points
az12_array = np.repeat(az12,points)
#Creating the distance array used to create the points.
dist_array_repeat = np.repeat(distances,points) * points_space
#Creating sampled points in between the given lat and lon coordinates for
#way points.
lon, lat, backaz = g.fwd(lon0_start_array, lat0_start_array, 
                         az12_array, dist_array_repeat) 
#The end point is not returned from the above calculation
```

Thanks for, from what I have seen, the most efficient package for what I want to do.

**Comment by cffk** [comment]
Try using the npts method for Geod.  This might speed things up.  However, I notice from
the code that the npts method itself could easily be speeded up.  I'll submit a separate issue
about this.

**Comment by AdInferos** [comment]
Hi cffk. 
Thanks. 
npts might be faster in itself compared to inv/fwd, but it does not handle an array input (a route with several waypoints) so I would have to loop through it, thus making the code significantly slower than the presented example.

**Comment by cffk** [comment]
OK.  FYI, the best-case timing for a calculation like this is about 0.35 us per waypoint.  This is based
on times for the C++ code (with all the looping done within compiled code) -- I expect the C timings to
be similar.

**Comment by AdInferos** [comment]
Ok, interesting. 
I just tested the above code snip. It results in roughly 0.7 * 10^(-6) seconds per produced point, so there is still some performance gain to achieve. If it is possible with a route with waypoints not on one great circle line.
Using npts by looping through the waypoints (as shown below), I get an  increase in computation time close to 100 times that of the above inv/fwd method (0.65*10^(-4)). 

```
for i in range(len(lon0)-1):
    lonlat_tuple = g.npts(lon0[i], lat[i], lon[i+1], lat[i+1], points[i], radians=False)
```


Thanks for all your help so fare.

**Comment by cffk** [comment]
I've submitted pull request #115 to give you that extra factor of 2.  Please try it.
There's another 20% improvement available, if you're willing to accept waypoints
which are only approximately evenly spaced (+/- 0.3%).  This involves specifying
the distance in terms of the arc length on the auxiliary sphere.  However, the
complexity of explaining what's going on and the modest performance gain (only
20%) means that it probably isn't appropriate for pyproj.  The C documentation
is [here](https://geographiclib.sourceforge.io/html/C/geodesic_8h.html#a5af4e14feebef638db75f2c87f14c088).

**Comment by AdInferos** [comment]
I have just updated my pyproj install. (checked it both using conda and pip)
I see an insignificant performance gain, but no where near a factor 2. 

It is the first time i actively update a package for after a pull request, so it might just be an older update I have installed.

**Comment by cffk** [comment]
The pull request hasn't been merged into pyproj.  So you would need to check out

git@github.com:cffk/pyproj.git
branch = geod-npts-fix

and go through the steps to build and install it.  Alternatively, sit tight and maybe
the pull request will be merged and included in the next release.

**Comment by micahcochran** [comment]
PR #115 was just merged for Geod.npts performance improvements.
